### ISSUE TITLE
[cloud_firestore] Transaction on iOS reports Unhandled Exception: PlatformException

### ISSUE THREAD

**Issue by matthew-carroll** [FirebaseExtended/flutterfire#1758]
**Describe the bug**
Reading and then writing a value in a transaction results in the following error:

[VERBOSE-2:ui_dart_state.cc(157)] Unhandled Exception: PlatformException(3, Every document read in a transaction must also be written in that transaction., null)

**To Reproduce**
```
final serializedEntry = // a map of values to add to an array

final DocumentReference docRef =
          Firestore.instance.collection('some_collection').document('some_document');

Firestore.instance.runTransaction((Transaction transaction) async {
      // Read the current snapshot for a document.
      DocumentSnapshot snapshot = await transaction.get(docRef);

      // Add an item to an array called "entries".
      List<dynamic> entries = List(snapshot.data['entries']);
      entries.add(serializedEntry);

      // Write the updated entries array back to firestore.
      await transaction.update(docRef, {
        'entries': entries,
      });
});
```

**Expected behavior**
The "entries" array in firestore is updated and the transaction function returns without error.

**Additional context**
I'm using v0.13.0+1 of the cloud_firestore plugin.

I'm running this code on a physical iPhone 7.

If I never reference the `snapshot` variable in the above sample code, the error does not occur. Therefore, at first blush it seems that accessing the `data` of the snapshot sets some kind of flag that leads to this error. For example, the following version of the repro code does *not* cause the error:

```
final serializedEntry = // a map of values to add to an array

final DocumentReference docRef =
          Firestore.instance.collection('some_collection').document('some_document');

Firestore.instance.runTransaction((Transaction transaction) async {
      // Read the current snapshot for a document.
      DocumentSnapshot snapshot = await transaction.get(docRef);

      // Write the updated entries array back to firestore.
      await transaction.update(docRef, {
        'entries': [serializedEntry],
      });
});
```

**Comment by bpaul7101** [comment]
awaiting the runTransaction() ... fixes issues like this for me ... and also changing the update to use the READ value of the snapshot, i.e

```
  await transaction.update(snapshot.reference, {
        'entries': [serializedEntry],
      });
```

**Comment by Salakar** [comment]
Hey all, as part of our on-going work for #2582, this has been resolved in our Firebase Firestore rework (#2913) - which has now been merged into master. We'll look at publishing some prereleases in the next few days. Thank you
